### ISSUE TITLE
Discussion: cloud workflows and the distribution effect

### ISSUE THREAD

**Issue by eiriktsarpalis** [mbraceproject/MBrace.Core#9]
Drawing on the discussion started in this [issue](https://github.com/mbraceproject/MBrace.Azure/issues/7), I would like to share a few thoughts on the programming model.

As you may know, cloud workflows are used in every aspect of the MBrace API, from parallel combinators to store operations. For instance, the `ICloudDisposable` interface has the following signature:
```fsharp
type ICloudDisposable =
    abstract Dispose : unit -> Cloud<unit>
```
An interesting question that arises here is, how one can know if a dispose implementation does not introduce distribution? While it makes sense that all primitive store operations should not introduce distribution, this cannot be guaranteed by their type signature. A workflow of type `Cloud<unit>` could either signify asynchronous store operation or it could contain a massively distributed computation. In other words, there is no way to statically detect if a workflow carries the distribution effect.

Currently, this is somewhat mitigated using the following primitive:
```fsharp 
Cloud.ToLocal : Cloud<'T> -> Cloud<'T>
```
This has the effect of evaluating the input workflow with thread pool parallelism semantics, thus giving a dynamic guarantee that the nested computation will never exit the current worker machine. It offers relative sanity, but is hard to reason about and does not work correctly in conjunction with forking operations, like `Cloud.StartChild`.

My proposal to amending this issue is to introduce two brands of computation expressions for MBrace workflows, for local and distributed computations. A distributed workflow can compose local workflows, but not the other way around. Store operations will be local workflows and the parallelism primitives will necessary return distributed workflows. This would allow to statically reason about the distribution effect, while potentially complicating the programming model.

I have created an [experimental branch](https://github.com/mbraceproject/MBrace.Core/tree/local-workflows) that attempts to develop these ideas:
[Workflow definitions](https://github.com/mbraceproject/MBrace.Core/blob/local-workflows/src/MBrace.Core/Continuation/Cloud.fs#L34)
[Builder declarations](https://github.com/mbraceproject/MBrace.Core/blob/local-workflows/src/MBrace.Core/Continuation/CloudBuilder.fs#L230)
[Store operations using local workflows](https://github.com/mbraceproject/MBrace.Core/blob/local-workflows/src/MBrace.Core/Store/CloudRef.fs#L97)
[Cloud.Parallel primitive](https://github.com/mbraceproject/MBrace.Core/blob/local-workflows/src/MBrace.Core/Runtime/RuntimeCombinators.fs#L115)

Thoughts?

**Comment by eiriktsarpalis** [comment]
To give a more concrete example, the proposed change in the programming model would essentially contain two types, `Local<'T>` and `Cloud<'T>`, which would both be subtypes of the abstract type `Workflow<'T>` (names are indicative and subject to revision, of course). Each of the two subtypes would come with distinct expression builders:
```fsharp
let readFile (reader : Stream -> 'T) (f : CloudFile) : Local<'T> = local {
    use! stream = CloudFile.ReadAsStream f
    return reader stream
}
```
and
```fsharp
let readParallel (reader : Stream -> 'T) (files : CloudFile []) : Cloud<'T []> = cloud {
    return! files |> Array.map (readFile reader) |> Cloud.Parallel
}
```
The suggested bind operation in local workflows will have type signature `Local<'T> -> ('T -> Local<'S>) -> Local<'S>`, whereas in cloud workflows it will carry the signature `Workflow<'T> -> ('T -> #Workflow<'S>) -> Cloud<'S>`. In other words this would mean that local workflows freely compose with other local workflows, whereas they cannot bind to cloud workflows. On the other hand, cloud workflows can freely bind to both local and cloud workflows. 

The `Cloud.Parallel` keyword will have type signature `seq<#Workflow<'T>> -> Cloud<'T []>`; since the primitive introduces distribution by definition, it will force 'elevation' of any workflow into a cloud workflow. At the same time, any use of `Cloud.Parallel` inside a `local` workflow will introduce a type error. For threadpool-based parallelism, a `Local.Parallel` primitive will be introduced, of type signature `seq<Local<'T>> -> Local<'T []>`. This will offer static assurance that the this local parallelism combinator cannot compose with child computations that introduce parallelism.

In other words, the proposed amendment to the programming model aspires to constrain distribution effects at compile time.

**Comment by palladin** [comment]
Given the fact that Cloud<_> and Local<_> are defined as
```fsharp
type Local<'T> = Workflow<Local, 'T>
type Cloud<'T> = Workflow<Distributed, 'T>
```
I was thinking that [CloudBuilder.Bind]
(https://github.com/mbraceproject/MBrace.Core/blob/local-workflows/src/MBrace.Core/Continuation/CloudBuilder.fs#L240) could be  something like
```fsharp
member __.Bind<'Ctx, 'T, 'S (f : Workflow<'Ctx, 'T>, g : 'T -> Workflow<Distributed, 'S>) : Workflow<Distributed, 'S>
```

**Comment by eiriktsarpalis** [comment]
Well, that's essentially identical to the current signature.

**Comment by palladin** [comment]
I'm referring to this (f : #Workflow<'T>)
https://github.com/mbraceproject/MBrace.Core/blob/local-workflows/src/MBrace.Core/Continuation/CloudBuilder.fs#L240
I think it is better for all types of bind to be of the same form Workflow<_, _>

**Comment by eiriktsarpalis** [comment]
I have a few updates on the proposed changes. I have pushed a [new branch](https://github.com/mbraceproject/MBrace.Core/tree/local-workflows-2) that proposes an alternative representation for local workflows.

The [initial branch](https://github.com/mbraceproject/MBrace.Core/tree/local-workflows) uses a phantom type approach
```fsharp
[<Abstract>] type Workflow<'T>

type Workflow<'Ctx, 'T when 'Ctx :> SchedulingContext> = inherit Workflow<'T>

type Local<'T> = Workflow<Local, 'T>
type Cloud<'T> = Workflow<Cloud, 'T>
```
This approach comes with the drawback that phantom types may be confusing to novice users. Type errors when doing incompatible compositions may also be difficult to read. What's more, the type alias declarations are only visible from F# code; a C# library would have to define separate types of wrappers which would possibly hurt interoperability. 

The new branch uses the following, more straightforward approach:
```fsharp
[<Abstract>] type Workflow<'T>

type Local<'T> = inherit Workflow<'T>
type Cloud<'T> = inherit Workflow<'T>
```
This has better readability, but restricts scheduling context definition to the core library. It also does not allow context polymorphism. For instance, the combinator
```fsharp 
Cloud.Catch : Workflow<'Ctx, 'T> -> Workflow<'Ctx, Choice<'T, exn>>
```
has to be rewritten to two concrete implementations:
```fsharp
Local.Catch : Local<'T> -> Local<Choice<'T,exn>>
Cloud.Catch : Cloud<'T> -> Cloud<Choice<'T,exn>>
```
It is not clear to me which of the two approaches is simplest.

**Comment by isaacabraham** [comment]
I must admit, I had until now no clue what the term "phantom types" even referred to. Now I think I see what you're saying - basically, in the first implementation both Cloud and Local are both really the same type, but at compile time you have the notion of two distinct types. In this way you can write operations that can apply across both simply by targeting the "real" type, or just one of them by using one of the two phantom types. OK. So this probably has lots of benefits to the implementation I would imagine in terms of code reuse and simplification? So from an F# perspective it's only positives I guess? Could you give examples of "bad" error messages caused from this?

Alternative is to create two concrete types with an inheritance chain - with means from a runtime perspective we get two types but there's a duplication of code etc..

What I like is that both approaches tend to suggest that the beginner can simply use Cloud everywhere and not worry about locality, and as their understanding improves they can adopt Local for efficiency etc.. The C# point is also valid and is important. What's going to be the cost in terms of the wrapper API if you're using phantom types? Is it going to be massively more costly to implement the wrapper then this is something you should think very carefully about. If it's only a little bit more costly to do, I would tend to go down the phantom types approach.

**Comment by eiriktsarpalis** [comment]
Well, it does result in some code duplication but this really only affects the internals of the core library.

In terms of error messages, given the following piece of code:
```fsharp
local {
    let! x = Cloud.Parallel [ cloud { return 42 }]
    return x
}
```
The former approach would give:
```
Error	1	Type mismatch. Expecting a
    Local<'a>    
but given a
    Cloud<int []>    
The type 'Context.Local' does not match the type 'Context.Cloud'
```
whereas the latter woud give:
```
Error	1	This expression was expected to have type
    Local<'a>    
but here has type
    Cloud<int []>
```
The first case is not terrible, but the last line might be a bit confusing.

As regards the C# api, I don't think it's so much an issue of performance as one of composability. It should be possible to compose workflows defined in C# and F# libraries interchangeably, without having to wrap or unwrap types.

**Comment by isaacabraham** [comment]
OK. The first error isn't actually that bad really, but yes, the second is more what I would be used to. Regarding the C# API, I was more meaning the issue of effort to write wrappers etc. rather than performance.

Yes, it's a nice goal to compose workflows in either language without wrapping but this might also limit your options in terms of what types you expose in F# because you're thinking about how they look to C# e.g. option types, discriminated unions etc.. Wrapping might be the only course to achieve an idiomatic C# library. I believe that Akka .NET have done the same thing albeit the other way around. Orleans on the other hand....

I've not had a look at the C# API yet though so I might be spectacularly off course here :-)

**Comment by eiriktsarpalis** [comment]
Yes, I think you're right. It still seems to me that phantom types would make a cool F# snippet, but feel a bit strange when used as the centerpiece of a large framework.

**Comment by isaacabraham** [comment]
If the shoe fits... :-) Until we're at 1.0 there's nothing wrong with putting stuff in and removing it IMHO. Particularly if this is going to speed up development time - as we approach a stage when you feel that the core feature set is ready, that's when you might want to review this stuff.

Again, IMHO.

**Comment by dsyme** [comment]
@eiriktsarpalis  - Can Local<T> workflows wait on cloud I/O actions? e.g. reads from CloudRef, or Cloud.Log?

**Comment by eiriktsarpalis** [comment]
Yes. Only Cloud.Parallel/Choice/StartAsCloudTask require explicit use of cloud {}

**Comment by dsyme** [comment]
So what happens if ``Local<T>`` and ``Cloud<T>`` are completely separate types and ``Workflow<T>`` is completely hidden?  It sounds like there is very little code that is generic over ``Workflow<T>``?

**Comment by eiriktsarpalis** [comment]
Most combinators that return Cloud<> take Workflow<> as input. For instance
```fsharp
Cloud.Parallel : seq<#Workflow<'T>> -> Cloud<'T>
```
and
```fsharp
cloud.Bind : Workflow<'T> -> ('T -> Cloud<'S>) -> Cloud<'S>
```
Having a base type essentially simplifies APIs that take any workflow as input.

**Comment by dsyme** [comment]
OK, I see. Tricky choice!

**Comment by eiriktsarpalis** [comment]
Plus, it would be impossible to have this otherwise reasonable expression
```fsharp
Cloud.Parallel [ cloud { return 1 } ; local { return 2 } ]
```

**Comment by isaacabraham** [comment]
@eiriktsarpalis Hmmm. Does that mean "run these two workflows in parallel across the cluster - first one is a cloud workflow, ok. second one is a local workflow, promote it to a cloud workflow"?

**Comment by eiriktsarpalis** [comment]
@isaacabraham Not exactly. It means that both types can be run remotely, however the first one may encapsulate further distribution, while the second comes with a static guarantee that it will be constrained to a single machine if run. For a more illustrative example:
```fsharp
Cloud.Parallel [ Cloud.Parallel ... ; local { return! CloudRef.New 42 }]
```
The first argument and left argument are of type cloud since their execution *might* span multiple machines, while the right argument will not. Attempting to invert things, for instance
```fsharp
Local.Parallel [ Cloud.Parallel ... ]
```
will result in a type error. Having type local is a guarantee that no distribution will occur as far as MBrace is concerned inside that particular workflow. However, having type cloud does not necessarily imply distribution, for instance  `cloud { return 1 }`.

**Comment by eiriktsarpalis** [comment]
This addition is merely a type-level safety blanket. It does not affect the underlying implementation or execution model. It has actually helped uncover a few bugs, in some cases where distribution had been accidentally introduced in store APIs. Overall it has increased my confidence in working with the model, my only hope is that it has not been rendered incomprehensible to outsiders in the process.

**Comment by isaacabraham** [comment]
OK, yes, got it. It's not incomprehensible IMHO, but it needs to be introduced carefully within tutorials etc. so that users understand how to reason about local { } versus cloud { } - it's definitely not a beginner "feature". The beauty is that because everything works under the Cloud expr., there's no need to introduce it immediately IMHO.

**Comment by palladin** [comment]
Based on my experience, is very helpful to have the static guarantee but  also the visual cues (cloud {} vs local {}) which is necessary to understand the dynamics of you code.

**Comment by eiriktsarpalis** [comment]
Exactly! The novice need not even notice that there is a local expression builder. It could be introduced further on, in a 'constraining distribution' chapter.

**Comment by isaacabraham** [comment]
@eiriktsarpalis the only thing I would suggest you think about is namespacing of this expr. - if cloud { } is at the top level of MBrace etc., I would suggest putting local { } in a sub-namespace. The novice should (ideally) be able hit "." in intellisense and just see the constructs that would be useful to him for 80% of his or her needs etc.

**Comment by eiriktsarpalis** [comment]
Great! Let's see if we can settle the phantom type dispute then and merge the 'correct' branch to master. @dsyme ?

**Comment by dsyme** [comment]
The final position seems reasonable. This is the sort of addition which you have to be prepared to unroll if usability evidence indicates that's necessary, or if it collides with some other concern, but what you've got is worth running with for now.

**Comment by dsyme** [comment]
Some feedback on this design...

I was teaching fresh users for the first time today with this enabled (experienced F# programmers).  

* The presence of "Workflow" and "Local" was definitely noticed in the learning process - I normally explain combinators like ``Cloud.Parallel`` simply by showing their type, but now I had to explain what ``Workflow`` was too.  That definitely made things more complex to teach

* I was asked "why not use ``Async<T>`` for everything. I explained that it's because ``Async<T>`` values can't be serialized or would have different semantics w.r.t. shared memory.  I don't know if this is the best explanation or not, but it is an interesting question.  I think it becomes particularly interesting now we have three levels of workflows (Async --> Local --> Cloud)  - people will definitely ask "why so many" and I think it may impact adoption, especially in independent learning scenarios (in tutorials you can get people over this hump).

I still think the additional safety is worth it, but do wonder why "Local" and "Async" aren't the same thing.

**Comment by eiriktsarpalis** [comment]
Good question. Let me first clarify that Async values are perfectly serializable and can be composed/executed in MBrace workflows with the expected shared memory semantics. 

In fact, Async used to play the role of Local until very recently. However, Async lacks the dependency resolution mechanism used by MBrace workflows. For example:
```fsharp
type CloudFile = { Path : string }
    member file.ReadAllLines() = local {
        // resolve file store implementation particular to current execution context;
        // in azure it would be a blob store client bound to the user's connection string
        let! fileStore = Workflow.GetResource<ICloudFileStore> ()
        return fileStore.ReadAllLines(file.Path)
    }
```
Dependencies are threaded through the workflow in a reader monad fashion. The CloudFile instance contains only its path and connects to it using dependencies from the execution context. It would be impossible to do the same with Async without somehow encapsulating store logic.

With the update to the programming model, there are many things that need to brushed up. For example, the `Cloud.OfAsync : Async<'T> -> Local<'T>` primitive has a type signature that seems to contradict its name. We have kept it like this out of backwards compatibility for now, but we would love to have more feedback on how to improve things.

**Comment by palladin** [comment]
I think that maybe the phantom type design is more clear for experienced and "maybe" to newbies as well.
One possible simplification is to rename Workflow to Cloud<'Ctx, 'T>
Some examples:
Cloud.OfAsync : Async<'T> -> Cloud<Local, 'T>
Cloud.Parallel : seq<Cloud<'Ctx, 'T>> -> Cloud<Distrib, 'T []>
CloudBuilder.Bind : Cloud<'Ctx, 'T> -> ('T -> Cloud<Distrib, 'R>) -> Cloud<Distrib, 'R>
LocalBuilder.Bind : Cloud<Local, T> -> ('T -> Cloud<Local, 'R>) -> Cloud<Local, 'R>

**Comment by eiriktsarpalis** [comment]
But you would still use the `Local<'T>` alias for `Cloud<Local, 'T>` right? If yes, then confusion persists. If not, you confuse even more by forcing people to explicitly work with phantom types.

**Comment by palladin** [comment]
My idea is to remove the type aliases, I think that with the phantom types exposed is more clear (to me at least).

**Comment by dsyme** [comment]
One concept is dropped (Workflow) at the expense of the phantom type parameter

    Cloud.Parallel : seq<Cloud<'Ctx, 'T>> -> Cloud< 'T []>

**Comment by eiriktsarpalis** [comment]
We would also need to have a `Cloud<'T>` base type (to replace Workflow<'T>). Not sure which is less confusing TBH.

**Comment by palladin** [comment]
The confusion arises because we hide the context. Maybe we can keep the Cloud<'T> as a type alias of Cloud<Distrib, 'T> and delete the Local<'T> completely.

**Comment by palladin** [comment]
For the scenario of non-uniform Cloud.Parallel
```
Cloud.Parallel [ cloud { return 1 } ; local { return 2 } ]
```
we can provide two cast operators
```
Cloud.ToLocal : Cloud<Distrib, 'T> -> Cloud<Local, T>
Cloud.AsDistrib : Cloud<Local, 'T> -> Cloud<Distrib, T> 
```
and we can keep one uniform Cloud.Parallel
```
Cloud.Parallel [ cloud { return 1 } ; Cloud.AsDistrib <| local { return 2 } ]
Cloud.Parallel : seq<Cloud<'Ctx, 'T>> -> Cloud<Distrib, 'T []>
```

**Comment by palladin** [comment]
Another possible idea is to have something like
```
type Context
type Distrib = inherit Context
type Local = inherit Context

type Cloud<'Ctx, 'T when 'Ctx :> Context>

Cloud.Parallel<'Ctx when 'Ctx :> Context> :  seq<Cloud<'Ctx, 'T>> -> Cloud<Distrib, 'T []>

let first : Cloud<Context, int> = cloud { return 1 }
let second : Cloud<Context, in> = local { return 2 }
Cloud.Parallel [  first; second ]
```

**Comment by dsyme** [comment]
FWIW I'm not at all sure the current situation is a problem - people did notice "Workflow" and "Local", and asked about "Async", but equally they understood my explanation and saw that the distinctions brought value.

The real test will come when 
* people try to learn from documentation rather than a tutorial
* people less experienced in teaching F# try to teach MBrace

In both these situations you will notice that the complexity of what can be communicated is lower than you expect, and that simplicity is a total virtue. To be successful, MBrace must be virally communicable as a simple-but-powerful big data technology  - it's not just whether we can teach it, but whether the people we teach can teach it.

**Comment by palladin** [comment]
Valid point. The teaching aspect is the critical one.

**Comment by eiriktsarpalis** [comment]
What if I made Cloud<> the base type and had Local<> be its subtype? We
could probably get rid of Workflow<> without changing much else.

On Friday, March 13, 2015, Nick Palladinos <notifications@github.com> wrote:

> Valid point. The teaching aspect is the critical one.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mbraceproject/MBrace.Core/issues/9#issuecomment-78690275>
> .
>


-- 
Sent from Gmail Mobile

**Comment by palladin** [comment]
Simplification is always a good thing, one less thing to worry about.

**Comment by dsyme** [comment]
But then a Local could be used where a Cloud is expected?

**Comment by eiriktsarpalis** [comment]
Yes, that is in fact expected behaviour, even now. It is the opposite situation that we are looking to avoid. 

Actually, my original planning was to do it like this, but then I went for the phantom type approach out of convenience. I complete forgot about this afterwards. The more I think about this, the more it makes sense to do it like this.

**Comment by dsyme** [comment]
Yes, sorry, brain fail :)

**Comment by eiriktsarpalis** [comment]
Ok, this refactoring worked out smoothly: https://github.com/eiriktsarpalis/MBrace.Core/commit/ddb77c29c65c055408b22291f560f269c42cb901

If everybody agrees, I'll proceed with a PR.

**Comment by eiriktsarpalis** [comment]
As an added bonus, this is now possible:
```fsharp
Cloud.Parallel [ cloud { return 1 } ; local { return 2 } :> _ ]
```

**Comment by palladin** [comment]
Yep, I think it works elegantly!

**Comment by dsyme** [comment]
Feedback time: 

First the very good :) I just took a non-PL F# user through MBrace - we concluded the session by running the scale-out comparison of feature extraction sets for 700,000 images over 200 Azure machines (currently running).  It was really fun to use the system in anger over data.  We'll write up more details in due course.

However, I do have to mention that both ``Local<_>`` and the ```#``` in ```#Cloud<_>``` in signatures are definitely noticed they add non-trivial learning burden for users.  It is really quite hard for people to grok why subtyping suddenly makes an appearance, especially since they are in the process of learning an already abstract notion like ```Cloud```.  

When someone knowledgeable is on hand to get them over this hurdle with confidence, it is ok.  But if they are on their own, I think most entry-level F# users are going to struggle with Cloud v. Local v. Async v. CloudFlow.  It's just hard for people to learn these things and understand why so many new control constructs are needed.

So the feedback is similar to before: ``Local<_>`` reduces bugs,  I like it. I'm not saying we should remove it.  But ``Local<_>`` adds a non-trivial burden to the learning curve, even for F# programmers, and I'm concerned about that. People accept an explanation when it is given but if left to their own devices they might not like it.  It's quite possible people will reject MBrace as "too complicated" when they see this alone.

One option may be to keep pushing ``Local<_>`` and ``#Cloud<_>`` to the fringe so you don't really notice it until you need it.  The specific place where "#Cloud<_>" gets noticed is in the signature to ``Cloud.Parallel`` and ``Cloud.Choice``.  To be honest we could just make this accept type ``Cloud<_>`` and have a ``Local.Parallel`` and ``Local.Choice`` as well.  

The specific place where ``Local<_>`` gets noticed is in the return type to ``CloudCell.New<_>``.  This is harder to hide.

**Comment by isaacabraham** [comment]
Where is there any details on CloudFlow?

**Comment by dsyme** [comment]
CloudFlow is the new name for CloudStream

**Comment by eiriktsarpalis** [comment]
Perhaps the root of problem lies in the naming of `Local<_>`. Just by name, it would be impossible to guess that this could be a subtype of `Cloud<_>`. What if we called it `LocalCloud<_>` or `Cloudlet<_>` or `Vapor<_>`? We would probably have to rename `local { }` which I have grown incredibly fond of.

**Comment by dsyme** [comment]
That's certainly one issue, though the root of the problem is simply the existence of more concepts for people to learn rather than less. 2 concepts are just harder to learn than 1 (while equally, 2 concepts may catch more bugs than 1 concept).  

I assume a C# version of all this would likely just recycle ``Task<T> = Local<T> = Cloud<T>`` and be called "task execution", no matter what the correctness problems with that.  But people would probably flock to it - partly because it's C# and partly because it's perceived to be simpler. Then a few people would realize the problems with it and start to refine the types used, but most wouldn't....

Another POV is that the Local v. Cloud distinction may be of more use to us as implementors (and also to power uses) than it is to 90% of potential users.  If that were deemed to be the case then perhaps it's something that could be present in debug mode (when building the library), and absent in release mode, though that's somewhat tricky to arrange.

Still another POV is that you could opt-in to a "stricter" API using "open MBrace.MoreExactTypes" or something, and without that you just get ``Cloud<T>``

But I'm really not sure if we need a change here - I'm really just giving feedback from 1:1 learning sessions.  We'll learn more as we teach it more.  However I've definitely noticed that some F# frameworks suffer from "too much type-level complexity" and "too many things to learn", and I believes it can hinder their adoption.

For the names above - I think local is better than those - people learn "local" in juxtaposition to "cloud" so they get the difference.

**Comment by dsyme** [comment]
Perhaps an alternative is this naming:

    type Cloud<'T>
    type Cloud<'T, 'Where> : Cloud<T>
    type Local  // a tag type

So what we call ``Local`` today is always seen as ``Cloud<T, Local>``. Then the average user only ever sees

     Cloud<T>
     Cloud.Parallel: seq<Cloud<T>>  -> Cloud<T[]>

And the power users sees:

    cloudLocal { ... }
    Cloud<T,Local>
    CloudLocal.Parallel: seq<Cloud<T,Local>>  -> Cloud<T[],Local>

For the power user the machinery is at least reusable if they choose to define their own tags, with the added bonus that combinators like CloudLocal.Parallel also preserve localness.

**Comment by palladin** [comment]
The structure as it is right now works beautifully, maybe we need to rename Local<'T> to CloudLocal<'T> in order to give it some context. But local {} is too beautiful to change it.

**Comment by isaacabraham** [comment]
@palladin @eiriktsarpalis: I personally really like the ```local { }``` abstraction. Once I "got" it, it allowed me to more easily reason about my code.

However, I'd also strongly recommend acting on this feedback, even if it's negative or not necessarily in line with what was hoped for. If people are struggling now - and I assume from Don that most of the individuals he's coaching are MSR people - then the average developer will probably experience the same. Most of those individuals won't have an F# expert sitting by their side either. Some of those might well just give up if they get stuck on something like this.

So whilst I think I'm in agreement with you regarding the effectiveness of ```local { }``` I'm also worried about users struggling to get up to speed with the different abstractions - something I've seen as well.

**Comment by eiriktsarpalis** [comment]
I'm not sure if renaming `Local<'T>` would somehow ease understanding for novices. Combinators that explicitly require local as arguments would type error if supplied with cloud. So if the goal is to delay introduction of the concept, I find it unlikely that it will be achieved in this way.

Having played with local workflows quite I bit, I can say that the local/cloud duality is a central point of the programming model as it is. Perhaps it would make sense to promote this distinction from the very beginning in tutorials.

**Comment by dsyme** [comment]
Here's an overview of the puppy image real-world scenario from yesterday. Basically the work came down to transporting 2GB (N * M * Size) of ``(string * int[])`` data to the cloud, running (N*N*M) Set.ofArray/Set.intersection, operations, checking if the intersection size was greater than some threshold (indicating similar or duplicate images), and returning strings (indicating the duplicate image names).  It turned out it could be composed to M independent jobs, each running at most 12 hours, so we used a 150 machine cluster to do it   There were probably a whole lot of optimizations we could have done (bit sets etc.), but we didn't need to bother.

In this scenario the final solution just ended up using nothing but ``cloud { ... }``, ``CreateProcess``, ``AwaitResult()`` and that's all.  Any data transport to the cloud was implicit in the ``cloud { ... }`` blocks.

We experimented with Cloud.Parallel but the overall size of the serialized job was too big, so we broke the work into the M independent CreateProcess calls (this is also why we were trying to parallelize calls to CreateProcess hence our bug report about that).  For a while we thought we might have to store the data in the cloud so we started to do M jobs doing CloudCell.New, but then we realized that storage was temporary and could be fused out.

Most of our actual work was preparing/shaping/trialing small trial jobs (e.g. M=3, N=3) to estimate how much compute and process-upload time we were going to need in total.

This scenario seems very typical of the "medium-data-plus-big-compute-in-the-cloud" scenario that MBrace will absolutely excel at.  For this scenario, the magic of MBrace is in REPL scripting and seamless-data-plus-code-transport-to-the-cloud.  The person I was working with was stunned by the ease and simplicity of that - what you've made with Vagrant + MBrace + Brisk is an amazing, simple,  exploratory, playful cloud scale-out programming environment.

I'll experiment with a PR for this idea: https://github.com/mbraceproject/MBrace.Core/issues/9#issuecomment-88632280, I'm fairly positive about this.

**Comment by eiriktsarpalis** [comment]
@dsyme Great, we would love to have his testimonial on the website once he's done.

**Comment by dsyme** [comment]
@eiriktsarpalis @palladin 

Reopening this old chestnut again...

Unfortunately I've had the feedback that "Learning local ... is really confusing" once again.

I'm still not sure the ``local { ... }`` v. ``cloud { ... }`` distinction is hitting the sweet spot for users (as opposed to combinatory-implementors). That is compared to alternatives like "everything is ``cloud { ... }``".  

First, the word "local" is still confusing people - is it "local to the worker" or "local to the client" or "local to a machine" or ...  People seem to be interpreting it as "local to the client" because of terminology like ``cluster.RunLocally``.  

Second, I'm just not sure that cloud v. local seems is a distinction that's _so_ important to the majority of users. People seem to be really, really confused by it and don't understand what it's giving them. The majority of uses of MBrace are where there is nothing but either cloud flows or "start lots of jobs and wait for them".

**Comment by dsyme** [comment]
Do you think it might be possible to somehow enable this distinction optionally, for authors of libraries like MBrace.Azure and MBrace.Flow, by opening extra namespaces? If that were possible it might deal with the problem.

**Comment by eiriktsarpalis** [comment]
I think this problem essentially boils down to the naming of the `Local<_>` type. The `local` expression builder as well as the `Local.*` methods can be easily hidden away in separate namespaces and never be noticed by novice users. The `Local<_>` type however is pervasive and cannot be ignored. Perhaps a good rename of this type would solve this ambiguity. We go for the phantom type approach or just call it `LocalCloud<_>`.

**Comment by palladin** [comment]
What if we remove the static typing of local {} and bring back dynamic typing-checking! 
Example
```
let mapCloud (f : 'T -> Cloud<'R>) (x : 'T) = cloud {
let! r = local (f x) // local : Cloud<'T> -> Cloud<'T>
}


mapCloud (fun x -> cloud {
                                let! y = Cloud.Parallel[] // exception boom runs in local context
                                return y
                                })
```

**Comment by eiriktsarpalis** [comment]
@palladin I'm strongly opposed to such an approach.

**Comment by dsyme** [comment]
:)  

I spiked two possible changes here: https://github.com/mbraceproject/MBrace.Core/compare/master...dsyme:fix-local

  - Abbreviate ``Local<T>`` = ``Cloud<T>`` in release mode, with ``mapLocal`` --> ``mapCloud``

  - Rename ``Local`` --> ``Cloud0`` and ``local`` --> ``cloud0``

and use the terminology "single-machine cloud workflow"

**Comment by isaacabraham** [comment]
Wasn't there a discussion about replacing local with async at some point as the semantics are somewhat similar?

An alternative is to go back to cloud { } but with some way of indicating that some cloud workflows only operate locally. I don't like this idea though.

I'm of the opinion that local { } has real value in being able to reason about your code and where it executes - which is one of the harder bits of Mbrace for beginners - and perhaps this problem can be solved by simply hiding local { } from higher namespaces and putting some xml comments for the beginner on local to say "just treat this as cloud if you're a beginner" :)

**Comment by palladin** [comment]
I like the terminology  "single-machine cloud workflow"

**Comment by dsyme** [comment]
The way I look at it is like this

|   |   CPU | Async I/O   |  Multi-threaded | Serialized   | Cloud I/O + Single machine |   Cloud I/O +multi machine | 
|:--|:------------:|:----:|:-----------:|:-----------:|:-----------:|:-----------:|:
| normal F# code |  x |  |   |  |
| ``async { ... }`` |  x | x  | x |  |  |
| ``cloud0 { ... }`` |  x | x  | x | x  | x | |
| ``cloud { ... }`` |  x | x  | x   |  x | x | x |

In this perspective an ``async0`` that is single-thread (can't start new child tasks in the thread pool, only supports StartImmediate) would make sense.

**Comment by dsyme** [comment]
@eiriktsarpalis - FWIW I gather from #117 (and previous discussions) that the ``cloud0``/``local`` row in the table above should not have an ``x`` in "Serialized semantics".  

This would mean that the only real difference between ``cloud0``/``local`` and ``async`` is that ``cloud0``/``local`` carries an extra computation-local data map - do I recall that correctly?

To put it another way, if you could edit the table above until it reflects what's accurate that would be great :)

**Comment by dsyme** [comment]
Copied from #117: 

@dsyme says:

I see, so the entire continuation is serialized, so

```
cloud {
    let v = ref 0 // or some non-serializable thing
    let! x = SomethingThatMayUseCloudParallel()
    ... reference to v ...
}
```

results in the possibility of invalid serialization of the continuation.

In this setting I still like the ``cloud0`` name - explained as "a ``cloud { ... }`` that starts precisely zero nested tasks, and no part of which gets serialized, and which can consequently use shared memory and unserializable in-memory objects safely"


@palladin says

> I was wondering what that cloud0 suffix zero means...

> " a cloud { ... } that starts precisely zero nested tasks, and no part of which gets serialized, and which can consequently use shared memory and unserializable in-memory objects safely" it is exactly the motivation behind local {}

**Comment by dsyme** [comment]
(I've edited the table above to reflect my understanding)

**Comment by palladin** [comment]
I think that "Cloud I/O + cancellation + multi machine" implies "Scheduling operations may serialize (dangerous to use shared mutable memory and unserializable objects)"

**Comment by dsyme** [comment]
columns consolidated

**Comment by eiriktsarpalis** [comment]
Slightly related, the master branch of MBrace.Core includes support for local execution that emulates the effects of distribution. See [this example](https://github.com/mbraceproject/MBrace.Core/blob/master/samples/MBrace.Thespian/test.fsx#L43).

**Comment by dsyme** [comment]
I took another less dramatic attempt to make progress on this issue here: https://github.com/mbraceproject/MBrace.Core/pull/119

This tries to replace the language "Remote" by "Cloud" and "Locally" by "Client".  Minimally I think we have to be really careful not to use "Local" to mean "Client"

**Comment by isaacabraham** [comment]
Given today's discussion, can this issue be closed as well?

**Comment by eiriktsarpalis** [comment]
Yep, closing.
